fri/src/prover/monolith/mod.rs

```diff
@@ -152,16 +156,17 @@ where
         let mut domain_size = self.layers[0].evaluations.len();
         let folding_factor = self.options.folding_factor();
 
-        // for all trees, except the last one, record tree root, authentication paths
-        // to row evaluations, and values for row evaluations
+        // for all FRI layers, except the last one, record tree root, determine a set of query
+        // positions, and query the layer at these positions.
         let mut layers = Vec::with_capacity(self.layers.len());
         for i in 0..self.layers.len() - 1 {
             positions = fold_positions(&positions, domain_size, folding_factor);
 
+            // sort of a static dispatch for folding_factor parameter
             let proof_layer = match folding_factor {
-                4 => build_proof_layer::<B, E, H, 4>(&self.layers[i], &positions),
-                8 => build_proof_layer::<B, E, H, 8>(&self.layers[i], &positions),
-                16 => build_proof_layer::<B, E, H, 16>(&self.layers[i], &positions),
+                4 => query_layer::<B, E, H, 4>(&self.layers[i], &positions),
+                8 => query_layer::<B, E, H, 8>(&self.layers[i], &positions),
+                16 => query_layer::<B, E, H, 16>(&self.layers[i], &positions),
                 _ => unimplemented!("folding factor {} is not supported", folding_factor),
```

Comment:
Is there a reason this is not generic? Or even, say, generic up to a power of 2? Is it because the hash function we use here only supports enough bits to fit up to 16 elements per hash? Could we replace it with a different function, since you use the type param H here anyway?

Reply:
This is my way of getting around the need to translate the value of the folding factor which comes from the `options` struct (and is known only at runtime) and a value of the folding factor used by `query_layer` function (which must be known at compile time). 

The only other more generic way of doing it (that I can think of) is to add folding factor as a generic type parameter to the `options` struct - but I'm trying to avoid doing this for a number of reason.